homeassistant/components/ring/sensor.py

```diff
@@ -59,183 +62,170 @@ async def async_setup_entry(
-        description: RingSensorEntityDescription,
+        description: RingSensorEntityDescription[RingGeneric],
     ) -> None:
         """Initialize a sensor for Ring device."""
         super().__init__(device, coordinator)
         self.entity_description = description
         self._attr_unique_id = f"{device.id}-{description.key}"
-
-    @property
-    def native_value(self):
-        """Return the state of the sensor."""
-        sensor_type = self.entity_description.key
-        if sensor_type == "volume":
-            return self._device.volume
-        if sensor_type == "doorbell_volume":
-            return self._device.doorbell_volume
-        if sensor_type == "mic_volume":
-            return self._device.mic_volume
-        if sensor_type == "voice_volume":
-            return self._device.voice_volume
-
-        if sensor_type == "battery":
-            return self._device.battery_life
-
-
-class HealthDataRingSensor(RingSensor):
-    """Ring sensor that relies on health data."""
-
-    # These sensors are data hungry and not useful. Disable by default.
-    _attr_entity_registry_enabled_default = False
-
-    @property
-    def native_value(self):
-        """Return the state of the sensor."""
-        sensor_type = self.entity_description.key
-        if sensor_type == "wifi_signal_category":
-            return self._device.wifi_signal_category
-
-        if sensor_type == "wifi_signal_strength":
-            return self._device.wifi_signal_strength
-
-
-class HistoryRingSensor(RingSensor):
-    """Ring sensor that relies on history data."""
-
-    _latest_event: dict[str, Any] | None = None
+        self._attr_entity_registry_enabled_default = (
+            description.entity_registry_enabled_default
+        )
+        self._attr_native_value = self.entity_description.value_fn(self._device)
 
     @callback
-    def _handle_coordinator_update(self):
+    def _handle_coordinator_update(self) -> None:
         """Call update method."""
-        if not (history_data := self._get_coordinator_history()):
-            return
-
-        kind = self.entity_description.kind
-        found = None
-        if kind is None:
-            found = history_data[0]
-        else:
-            for entry in history_data:
-                if entry["kind"] == kind:
-                    found = entry
-                    break
-
-        if not found:
-            return
-
-        self._latest_event = found
-        super()._handle_coordinator_update()
-
-    @property
-    def native_value(self):
-        """Return the state of the sensor."""
-        if self._latest_event is None:
-            return None
 
-        return self._latest_event["created_at"]
-
-    @property
-    def extra_state_attributes(self):
-        """Return the state attributes."""
-        attrs = super().extra_state_attributes
+        self._device = self._get_coordinator_data().get_device(
+            self._device.device_api_id
+        )
```

Comment:
```suggestion
        self._device = cast(
            _RingGenericT,
            self._get_coordinator_data().get_device(self._device.device_api_id),
        )
```
This is a limitation from the library. `get_device` always returns `RingGeneric`. Ideally it would return the correct subclass for a specific `device_api_id`.

_Possible, but that is advanced typing territory. Message me on discord if you're interested. Then we can see if we're able to get this to work together._

Reply:
Am very interested, have reached out.